### PR TITLE
Fixed  a performance problem

### PR DESCRIPTION
**Pull request by djbuch**
See explainations here : https://github.com/PrestaShop/ps_crossselling/pull/5

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Correct performance problems see : https://github.com/PrestaShop/ps_crossselling/pull/5
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20916)

### CODE REVIEW

File: classes/order/OrderDetail.php [hunk 1]
```diff
@@ -817,7 +817,7 @@ public static function getCrossSells($id_product, $id_lang, $limit = 12)
         SELECT o.id_order
         FROM ' . _DB_PREFIX_ . 'orders o
         LEFT JOIN ' . _DB_PREFIX_ . 'order_detail od ON (od.id_order = o.id_order)
-        WHERE o.valid = 1 AND od.product_id = ' . (int) $id_product);
+        WHERE o.valid = 1 AND od.product_id = ' . (int) $id_product . ' ORDER BY RAND() LIMIT ' . (int) $limit*4);
```
Comment (Progi1984) [inline classes/order/OrderDetail.php]:
```suggestion
        WHERE o.valid = 1 AND od.product_id = ' . (int) $id_product . ' ORDER BY RAND() LIMIT ' . (int) $limit * 4);
```

**Review by Progi1984 — COMMENTED** [review]
A small CS error

File: classes/order/OrderDetail.php (hunk 1 above)
Comment (djbuch) [reply classes/order/OrderDetail.php]: Yes saw it just now

**Review by Progi1984 — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Should be linked to an issue ;)

**Comment by djbuch** [comment]
Well I don't have any idea know but just saw the same code being used here so I made this PR, the issue is the same, and I think that ps_crossseling would need a refacto to use this function instead of using duplicated code, but right I don't have more time for this.

Regards

**Comment by PierreRambaud** [comment]
> Well I don't have any idea know but just saw the same code being used here so I made this PR, the issue is the same, and I think that ps_crossseling would need a refacto to use this function instead of using duplicated code, but right I don't have more time for this.
> 
> Regards

The problem here is this function is returning all cross sells products at the beginning to build an SQL request, if we accept your change this part is now useless https://github.com/PrestaShop/PrestaShop/pull/20916/files#diff-706c99b0aadf150c801cec8a83a90cc1R825 and the result will not be the good one.

To improve performance, maybe we should let MySQL do the job and use a subrequest instead of a built id list here https://github.com/PrestaShop/PrestaShop/pull/20916/files#diff-706c99b0aadf150c801cec8a83a90cc1R841

**Comment by PierreRambaud** [comment]
For more details, here a new diff which is improving SQL performance:

```
diff --git a/classes/order/OrderDetail.php b/classes/order/OrderDetail.php
index 8dd7c67132..90d4795c82 100644
--- a/classes/order/OrderDetail.php
+++ b/classes/order/OrderDetail.php
@@ -813,40 +813,37 @@ class OrderDetailCore extends ObjectModel
             $front = false;
         }
 
-        $orders = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS('
-        SELECT o.id_order
-        FROM ' . _DB_PREFIX_ . 'orders o
-        LEFT JOIN ' . _DB_PREFIX_ . 'order_detail od ON (od.id_order = o.id_order)
-        WHERE o.valid = 1 AND od.product_id = ' . (int) $id_product);
-
-        if (count($orders)) {
-            $list = '';
-            foreach ($orders as $order) {
-                $list .= (int) $order['id_order'] . ',';
-            }
-            $list = rtrim($list, ',');
-
-            $order_products = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS('
-                SELECT DISTINCT od.product_id, p.id_product, pl.name, pl.link_rewrite, p.reference, i.id_image, product_shop.show_price,
-                cl.link_rewrite category, p.ean13, p.isbn, p.out_of_stock, p.id_category_default ' . (Combination::isFeatureActive() ? ', IFNULL(product_attribute_shop.id_product_attribute,0) id_product_attribute' : '') . '
-                FROM ' . _DB_PREFIX_ . 'order_detail od
-                LEFT JOIN ' . _DB_PREFIX_ . 'product p ON (p.id_product = od.product_id)
-                ' . Shop::addSqlAssociation('product', 'p') .
-                (Combination::isFeatureActive() ? 'LEFT JOIN `' . _DB_PREFIX_ . 'product_attribute_shop` product_attribute_shop
-                ON (p.`id_product` = product_attribute_shop.`id_product` AND product_attribute_shop.`default_on` = 1 AND product_attribute_shop.id_shop=' . (int) Context::getContext()->shop->id . ')' : '') . '
-                LEFT JOIN ' . _DB_PREFIX_ . 'product_lang pl ON (pl.id_product = od.product_id' . Shop::addSqlRestrictionOnLang('pl') . ')
-                LEFT JOIN ' . _DB_PREFIX_ . 'category_lang cl ON (cl.id_category = product_shop.id_category_default' . Shop::addSqlRestrictionOnLang('cl') . ')
-                LEFT JOIN ' . _DB_PREFIX_ . 'image i ON (i.id_product = od.product_id)
-                ' . Shop::addSqlAssociation('image', 'i', true, 'image_shop.cover=1') . '
-                WHERE od.id_order IN (' . $list . ')
-                    AND pl.id_lang = ' . (int) $id_lang . '
-                    AND cl.id_lang = ' . (int) $id_lang . '
-                    AND od.product_id != ' . (int) $id_product . '
-                    AND product_shop.active = 1'
-                    . ($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '') . '
-                ORDER BY RAND()
-                LIMIT ' . (int) $limit . '
-            ', true, false);
+        $hasOrders = Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue(
+            'SELECT o.id_order ' .
+            'FROM ' . _DB_PREFIX_ . 'orders o ' .
+            'INNER JOIN ' . _DB_PREFIX_ . 'order_detail od ON (od.id_order = o.id_order) ' .
+            'WHERE o.valid = 1 AND od.product_id = ' . (int) $id_product
+        );
+
+        if ($hasOrders) {
+            $order_products = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS(
+                'SELECT DISTINCT od.product_id, p.id_product, pl.name, pl.link_rewrite, p.reference, i.id_image, product_shop.show_price, ' .
+                'cl.link_rewrite category, p.ean13, p.isbn, p.out_of_stock, p.id_category_default ' . (Combination::isFeatureActive() ? ', IFNULL(product_attribute_shop.id_product_attribute,0) id_product_attribute' : '') . ' ' .
+                'FROM ' . _DB_PREFIX_ . 'order_detail od ' .
+                'INNER JOIN ' . _DB_PREFIX_ . 'product p ON (p.id_product = od.product_id) ' .
+                Shop::addSqlAssociation('product', 'p') .
+                (Combination::isFeatureActive() ? 'INNER JOIN `' . _DB_PREFIX_ . 'product_attribute_shop` product_attribute_shop ' .
+                'ON (p.`id_product` = product_attribute_shop.`id_product` AND product_attribute_shop.`default_on` = 1 AND product_attribute_shop.id_shop=' . (int) Context::getContext()->shop->id . ')' : '') . ' ' .
+                'INNER JOIN ' . _DB_PREFIX_ . 'product_lang pl ON (pl.id_product = od.product_id' . Shop::addSqlRestrictionOnLang('pl') . ') ' .
+                'INNER JOIN ' . _DB_PREFIX_ . 'category_lang cl ON (cl.id_category = product_shop.id_category_default' . Shop::addSqlRestrictionOnLang('cl') . ') ' .
+                'INNER JOIN ' . _DB_PREFIX_ . 'orders o ON od.id_order = o.id_order AND o.valid = 1 ' .
+                'INNER JOIN ' . _DB_PREFIX_ . 'image i ON (i.id_product = od.product_id) ' .
+                Shop::addSqlAssociation('image', 'i', true, 'image_shop.cover=1') . ' ' .
+                'WHERE pl.id_lang = ' . (int) $id_lang . ' ' .
+                'AND cl.id_lang = ' . (int) $id_lang . ' ' .
+                'AND od.product_id != ' . (int) $id_product . ' ' .
+                'AND product_shop.active = 1' .
+                ($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '') . ' ' .
+                'ORDER BY RAND() ' .
+                'LIMIT ' . (int) $limit,
+                true,
+                false
+            );
 
             $tax_calc = Product::getTaxCalculationMethod();
             if (is_array($order_products)) {
```

You can apply this kind of change in ps_crosselling module too.

**Comment by PierreRambaud** [comment]
Hi @djbuch ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**